clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -400,12 +412,33 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
             int totalBytes = 0;
             int totalCount = 0;
             FetchResponse response = new FetchResponse(resp.responseBody());
+
+            // Look for OffsetOutOfRange error in fetchResponse
+            Map<TopicPartition, Long> offsetOutOfRangePartitions = new HashMap();
+            for (Map.Entry<TopicPartition, FetchResponse.PartitionData> entry : response.responseData().entrySet()) {
```

Comment:
I'm not sure that this is such a big problem. The records accumulate until they are retrieved and we always check that the record offset returned to the user is the one they expect. It seems like the worst thing that could happen is that the records continue to grow while we keep throwing out of range exceptions to the user. To address this, we should probably have KafkaConsumer.poll() check whether there are records to return before calling initFetches(). I think that would guarantee that no new fetches would be sent until all pending data is returned to the user. Do you think that would work?